pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/BaseChunkSVForwardIndexWriter.java

```diff
@@ -74,15 +71,13 @@ protected BaseChunkSVForwardIndexWriter(File file, ChunkCompressionType compress
       int numDocsPerChunk, int chunkSize, int sizeOfEntry, int version)
       throws IOException {
     Preconditions.checkArgument(version == DEFAULT_VERSION || version == CURRENT_VERSION);
-    _file = file;
-    _headerEntryChunkOffsetSize = getHeaderEntryChunkOffsetSize(version);
-    _dataOffset = headerSize(totalDocs, numDocsPerChunk, _headerEntryChunkOffsetSize);
     _chunkSize = chunkSize;
     _chunkCompressor = ChunkCompressorFactory.getCompressor(compressionType);
+    _headerEntryChunkOffsetSize = getHeaderEntryChunkOffsetSize(version);
+    _dataOffset = writeHeader(compressionType, totalDocs, numDocsPerChunk, sizeOfEntry, version);
     _chunkBuffer = ByteBuffer.allocateDirect(chunkSize);
-    _dataChannel = new RandomAccessFile(file, "rw").getChannel();
-    _header = _dataChannel.map(FileChannel.MapMode.READ_WRITE, 0, _dataOffset);
-    writeHeader(compressionType, totalDocs, numDocsPerChunk, sizeOfEntry, version);
+    _compressedBuffer = ByteBuffer.allocateDirect(chunkSize * 2);
```

Comment:
I suggest to use `ByteBuffer.allocateDirect(_chunkCompressor.maxCompressedSize(chunkSize))` instead because most compressors can produce a much tighter bound than 2x. E.g. LZ4 (default) produces 8240 for 8192 and 1052704 for 1MB.

Reply:
Good suggestion. Refactored.

Reply:
@sajjad-moradi  @richardstartin, I think 2x was being used to account for worst case where there can be negative compression which is a possibility (although unlikely) imo. Unless we are 100% sure that our compression codecs (LZ4, snappy, ZSTD) are never going to result in negative compression for any kind of data, using chunkSize is fine. Otherwise, I suggest moving it back to 2x (old code) or something more than chunkSize at least

Reply:
Each compression codec provides a bound for the worst case, i.e. when the data is incompressible and the result is larger than the input.

Reply:
Do we have these bounds available at run time (preferably as an API)? Otherwise, instead of trying to optimize to something we don't know, we should revert to code that we know works in production environment

Reply:
These bounds are specified by the compression codecs themselves. They take the size of the uncompressed input as a parameter and produce the largest possible compressed size for that input. This isn't trying to optimise something we don't know, but making proper use of compression libraries. Playing devil's advocate, I'm curious how you know 2x is large enough for _any_ data? Wouldn't you prefer to use the compression library's specified bounds?

Reply:
For posterity, here are links to the documentation of the functions we call in to via JNI:

`LZ4_compressBound` is described [here](https://fossies.org/linux/lz4/doc/lz4_manual.html)
`Snappy:: MaxCompressedLength ` is described [here](https://github.com/google/snappy/blob/master/snappy.h)
`ZSTD_compressBound` is described [here](https://zstd.docsforge.com/dev/api/ZSTD_compressBound/)

Reply:
I suggest making a change to use these if we really want to do better than 2x. Until then, we should switch back to old behavior of using 2x buffer size since it has stood the test of time and we know it works IMHO.

Reply:
@siddharthteotia What do you mean by negative compression? If you mean compressed size is more than the input size, then this maxCompressedSize function returns the upper bound for that - so we don't need to always go with 2x size.
BTW here's the max value for the input size of 1000 for different compressors:
```
pass through: 1000
snappy: 1198
zsandard: 1066
lz4: 1019
lz4 with length compressor: 1023
```